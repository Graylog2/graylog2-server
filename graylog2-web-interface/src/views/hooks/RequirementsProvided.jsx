// @flow strict
import * as React from 'react';
import { PluginStore } from 'graylog-web-plugin/plugin';

import MissingRequirements from 'views/components/views/MissingRequirements';
import type { ViewHook } from 'views/logic/hooks/ViewHook';
import type { Requirements } from 'views/logic/views/View';

const _missingRequirements = (requirements, requirementsProvided): Requirements => Object.entries(requirements)
  .filter(([require]) => !requirementsProvided.includes(require))
  .reduce((prev, [key, value]) => ({ ...prev, [key]: value }), {});

const RequirementsProvided: ViewHook = ({ view }) => {
  return new Promise((resolve, reject) => {
    const providedRequirements = PluginStore.exports('views.requires.provided');
    const missingRequirements = _missingRequirements(view.requires, providedRequirements);

    if (Object.keys(missingRequirements).length > 0) {
      const Component = () => <MissingRequirements view={view} missingRequirements={missingRequirements} />;

      return reject(Component);
    }

    return resolve(true);
  });
};

export default RequirementsProvided;
