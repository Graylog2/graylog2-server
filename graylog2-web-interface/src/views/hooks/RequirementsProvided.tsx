/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import { PluginStore } from 'graylog-web-plugin/plugin';

import MissingRequirements from 'views/components/views/MissingRequirements';
import type { ViewHook, ViewHookArguments } from 'views/logic/hooks/ViewHook';
import type { Requirements } from 'views/logic/views/View';

const _missingRequirements = (requirements: Requirements, requirementsProvided: string | string[]): Requirements => Object.entries(requirements)
  .filter(([require]) => !requirementsProvided.includes(require))
  .reduce((prev, [key, value]) => ({ ...prev, [key]: value }), {});

const RequirementsProvided: ViewHook = async ({ view, executionState }: ViewHookArguments) => {
  const providedRequirements = PluginStore.exports('views.requires.provided');
  const missingRequirements = _missingRequirements(view.requires, providedRequirements);

  if (Object.keys(missingRequirements).length > 0) {
    const Component = () => <MissingRequirements view={view} missingRequirements={missingRequirements} />;

    throw Component;
  }

  return [view, executionState];
};

export default RequirementsProvided;
