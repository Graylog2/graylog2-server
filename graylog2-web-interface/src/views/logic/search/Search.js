// @flow strict

import * as Immutable from 'immutable';
import ObjectID from 'bson-objectid';

import Query from '../queries/Query';
import Parameter from '../parameters/Parameter';
import type { ParameterJson } from '../parameters/Parameter';
import type { QueryJson } from '../queries/Query';

type SearchId = string;
type InternalState = {
  id: SearchId,
  queries: Immutable.Set<Query>,
  parameters: Immutable.Set<Parameter>,
};

export type SearchJson = {
  id: SearchId,
  queries: Array<QueryJson>,
  parameters: Array<ParameterJson>,
};

export type QuerySet = Immutable.Set<Query>;

export default class Search {
  _value: InternalState;

  constructor(id: SearchId, queries: (Array<Query> | QuerySet), parameters: Array<Parameter>) {
    this._value = { id, queries: Immutable.OrderedSet(queries), parameters: Immutable.Set(parameters) };
  }

  static create(): Search {
    // eslint-disable-next-line no-use-before-define
    return new Builder()
      .newId()
      .queries([])
      .parameters([])
      .build();
  }

  get id(): SearchId {
    return this._value.id;
  }

  get queries(): QuerySet {
    return this._value.queries;
  }

  get parameters(): Immutable.Set<Parameter> {
    return this._value.parameters;
  }

  get requiredParameters(): Immutable.Set<Parameter> {
    return this.parameters
      .filter((p) => (!p.optional && !p.defaultValue));
  }

  // eslint-disable-next-line no-use-before-define
  toBuilder(): Builder {
    const { id, queries, parameters } = this._value;

    // eslint-disable-next-line no-use-before-define
    return new Builder(Immutable.Map({ id, queries, parameters }));
  }

  toJSON(): SearchJson {
    const { id, queries, parameters } = this._value;

    return {
      id,
      queries: queries.toJS(),
      parameters: parameters.toJS(),
    };
  }

  static fromJSON(value: SearchJson): Search {
    const { id, parameters } = value;

    const queries = value.queries.map((q) => Query.fromJSON(q));

    return new Search(id, queries, parameters.map((p) => Parameter.fromJSON(p)));
  }

  // eslint-disable-next-line no-use-before-define
  static builder(): Builder {
    // eslint-disable-next-line no-use-before-define
    return new Builder();
  }
}

class Builder {
  value: Immutable.Map<string, *>;

  constructor(value: Immutable.Map<string, *> = Immutable.Map()) {
    this.value = value;
  }

  id(value: SearchId): Builder {
    return new Builder(this.value.set('id', value));
  }

  newId(): Builder {
    return this.id(ObjectID().toString());
  }

  queries(value: (Array<Query> | QuerySet)): Builder {
    return new Builder(this.value.set('queries', value));
  }

  parameters(value: Array<Parameter>): Builder {
    return new Builder(this.value.set('parameters', value));
  }

  build(): Search {
    const { id, queries, parameters } = this.value.toObject();

    return new Search(id, queries, parameters);
  }
}
