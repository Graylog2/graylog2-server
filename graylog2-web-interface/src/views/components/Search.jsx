// @flow strict
import * as React from 'react';
import { useCallback, useEffect } from 'react';
import PropTypes from 'prop-types';
import * as Immutable from 'immutable';
import styled, { css } from 'styled-components';
import type { StyledComponent } from 'styled-components';

import withLocation from 'routing/withLocation';
import type { Location } from 'routing/withLocation';
import connect from 'stores/connect';
import Footer from 'components/layout/Footer';
import AppContentGrid from 'components/layout/AppContentGrid';
import Sidebar from 'views/components/sidebar/Sidebar';
import WithSearchStatus from 'views/components/WithSearchStatus';
import SearchResult from 'views/components/SearchResult';
import type {
  SearchRefreshCondition,
  SearchRefreshConditionArguments,
} from 'views/logic/hooks/SearchRefreshCondition';
import { FieldTypesActions } from 'views/stores/FieldTypesStore';
import { SearchStore, SearchActions } from 'views/stores/SearchStore';
import { SearchExecutionStateStore } from 'views/stores/SearchExecutionStateStore';
import { SearchConfigActions, SearchConfigStore } from 'views/stores/SearchConfigStore';
import { SearchMetadataActions } from 'views/stores/SearchMetadataStore';
import { StreamsActions } from 'views/stores/StreamsStore';
import { ViewActions, ViewStore } from 'views/stores/ViewStore';
import HeaderElements from 'views/components/HeaderElements';
import QueryBarElements from 'views/components/QueryBarElements';
import WindowLeaveMessage from 'views/components/common/WindowLeaveMessage';
import IfDashboard from 'views/components/dashboard/IfDashboard';
import QueryBar from 'views/components/QueryBar';
import { FieldsOverview } from 'views/components/sidebar';
import DashboardSearchBar from 'views/components/DashboardSearchBar';
import SearchBar from 'views/components/SearchBar';
import CurrentViewTypeProvider from 'views/components/views/CurrentViewTypeProvider';
import IfSearch from 'views/components/search/IfSearch';
import IfInteractive from 'views/components/dashboard/IfInteractive';
import HighlightMessageInQuery from 'views/components/messagelist/HighlightMessageInQuery';
import { ViewMetadataStore } from 'views/stores/ViewMetadataStore';
import bindSearchParamsFromQuery from 'views/hooks/BindSearchParamsFromQuery';
import { useSyncWithQueryParameters } from 'views/hooks/SyncWithQueryParameters';
import { AdditionalContext } from 'views/logic/ActionContext';
import DefaultFieldTypesProvider from 'views/components/contexts/DefaultFieldTypesProvider';
import InteractiveContext from 'views/components/contexts/InteractiveContext';
import HighlightingRulesProvider from 'views/components/contexts/HighlightingRulesProvider';
import SearchPageLayoutProvider from 'views/components/contexts/SearchPageLayoutProvider';
import usePluginEntities from 'views/logic/usePluginEntities';

const GridContainer: StyledComponent<{ interactive: boolean }, void, HTMLDivElement> = styled.div`
  ${({ interactive }) => (interactive ? css`
    height: calc(100vh - 50px);
    display: flex;
    overflow: hidden;
    
    > *:nth-child(2) {
      flex-grow: 1;
    }
  ` : '')}
`;

const SearchArea: StyledComponent<{}, void, *> = styled(AppContentGrid)`
  height: 100%;
  z-index: 1;
  overflow-y: auto;
`;

const ConnectedSidebar = connect(
  Sidebar,
  { viewMetadata: ViewMetadataStore, searches: SearchStore, view: ViewStore },
  (props) => ({
    ...props,
    viewIsNew: props.view.isNew,
    queryId: props.viewMetadata.activeQuery,
    results: props.searches && props.searches.result ? props.searches.result.forId(props.viewMetadata.activeQuery) : undefined,
  }),
);

type Props = {
  location: Location,
};

const _searchRefreshConditionChain = (searchRefreshHooks, state: SearchRefreshConditionArguments) => {
  if (!searchRefreshHooks || searchRefreshHooks.length === 0) {
    return true;
  }

  return searchRefreshHooks.every((condition: SearchRefreshCondition) => condition(state));
};

const _refreshIfNotUndeclared = (searchRefreshHooks, executionState) => {
  const { view } = ViewStore.getInitialState();

  return SearchMetadataActions.parseSearch(view.search).then((searchMetadata) => {
    if (_searchRefreshConditionChain(searchRefreshHooks, { view, searchMetadata, executionState })) {
      FieldTypesActions.all();

      return SearchActions.execute(executionState);
    }

    return Promise.reject(searchMetadata);
  });
};

const SearchBarWithStatus = WithSearchStatus(SearchBar);
const DashboardSearchBarWithStatus = WithSearchStatus(DashboardSearchBar);

const ViewAdditionalContextProvider = connect(
  AdditionalContext.Provider,
  { view: ViewStore, configs: SearchConfigStore },
  ({ view, configs: { searchesClusterConfig } }) => ({ value: { view: view.view, analysisDisabledFields: searchesClusterConfig.analysis_disabled_fields } }),
);

const Search = ({ location }: Props) => {
  const { pathname, search } = location;
  const query = `${pathname}${search}`;
  const searchRefreshHooks: Array<SearchRefreshCondition> = usePluginEntities('views.hooks.searchRefresh');
  const refreshIfNotUndeclared = useCallback(
    () => _refreshIfNotUndeclared(searchRefreshHooks, SearchExecutionStateStore.getInitialState()),
    [searchRefreshHooks],
  );

  useEffect(() => {
    const { view } = ViewStore.getInitialState();

    bindSearchParamsFromQuery({ view, query: location.query, retry: () => Promise.resolve() });
  }, [location.query]);

  useEffect(() => {
    SearchConfigActions.refresh();

    StreamsActions.refresh();

    let storeListenersUnsubscribes = Immutable.List();

    refreshIfNotUndeclared().finally(() => {
      storeListenersUnsubscribes = storeListenersUnsubscribes
        .push(SearchActions.refresh.listen(refreshIfNotUndeclared))
        .push(ViewActions.search.completed.listen(refreshIfNotUndeclared));
    });

    // Returning cleanup function used when unmounting
    return () => { storeListenersUnsubscribes.forEach((unsubscribeFunc) => unsubscribeFunc()); };
  }, [refreshIfNotUndeclared]);

  useSyncWithQueryParameters(query);

  return (
    <CurrentViewTypeProvider>
      <IfInteractive>
        <IfDashboard>
          <WindowLeaveMessage />
        </IfDashboard>
      </IfInteractive>
      <InteractiveContext.Consumer>
        {(interactive) => (
          <SearchPageLayoutProvider>
            <DefaultFieldTypesProvider>
              <ViewAdditionalContextProvider>
                <HighlightingRulesProvider>
                  <GridContainer id="main-row" interactive={interactive}>
                    <IfInteractive>
                      <ConnectedSidebar>
                        <FieldsOverview />
                      </ConnectedSidebar>
                    </IfInteractive>
                    <SearchArea>
                      <IfInteractive>
                        <HeaderElements />
                        <IfDashboard>
                          <DashboardSearchBarWithStatus onExecute={refreshIfNotUndeclared} />
                        </IfDashboard>
                        <IfSearch>
                          <SearchBarWithStatus onExecute={refreshIfNotUndeclared} />
                        </IfSearch>

                        <QueryBarElements />

                        <IfDashboard>
                          <QueryBar />
                        </IfDashboard>
                      </IfInteractive>
                      <HighlightMessageInQuery>
                        <SearchResult />
                      </HighlightMessageInQuery>
                      <Footer />
                    </SearchArea>
                  </GridContainer>
                </HighlightingRulesProvider>
              </ViewAdditionalContextProvider>
            </DefaultFieldTypesProvider>
          </SearchPageLayoutProvider>
        )}
      </InteractiveContext.Consumer>
    </CurrentViewTypeProvider>
  );
};

Search.propTypes = {
  location: PropTypes.shape({
    query: PropTypes.object.isRequired,
    pathname: PropTypes.string.isRequired,
    search: PropTypes.string.isRequired,
  }).isRequired,
};

export default withLocation(Search);
