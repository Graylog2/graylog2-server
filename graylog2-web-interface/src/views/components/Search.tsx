/*
 * Copyright (C) 2020 Graylog, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the Server Side Public License, version 1,
 * as published by MongoDB, Inc.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * Server Side Public License for more details.
 *
 * You should have received a copy of the Server Side Public License
 * along with this program. If not, see
 * <http://www.mongodb.com/licensing/server-side-public-license>.
 */
import * as React from 'react';
import { useCallback, useEffect, useContext, useState } from 'react';
import * as Immutable from 'immutable';
import styled, { css } from 'styled-components';

import PageContentLayout from 'components/layout/PageContentLayout';
import withLocation from 'routing/withLocation';
import type { Location } from 'routing/withLocation';
import connect from 'stores/connect';
import Sidebar from 'views/components/sidebar/Sidebar';
import WithSearchStatus from 'views/components/WithSearchStatus';
import SearchResult from 'views/components/SearchResult';
import type {
  SearchRefreshCondition,
  SearchRefreshConditionArguments,
} from 'views/logic/hooks/SearchRefreshCondition';
import { FieldTypesActions } from 'views/stores/FieldTypesStore';
import { SearchStore, SearchActions } from 'views/stores/SearchStore';
import { SearchExecutionStateStore } from 'views/stores/SearchExecutionStateStore';
import { SearchConfigActions, SearchConfigStore } from 'views/stores/SearchConfigStore';
import { SearchMetadataActions } from 'views/stores/SearchMetadataStore';
import { StreamsActions } from 'views/stores/StreamsStore';
import { ViewActions, ViewStore } from 'views/stores/ViewStore';
import HeaderElements from 'views/components/HeaderElements';
import QueryBarElements from 'views/components/QueryBarElements';
import WindowLeaveMessage from 'views/components/common/WindowLeaveMessage';
import IfDashboard from 'views/components/dashboard/IfDashboard';
import QueryBar from 'views/components/QueryBar';
import { FieldsOverview } from 'views/components/sidebar';
import DashboardSearchBar from 'views/components/DashboardSearchBar';
import SearchBar from 'views/components/SearchBar';
import CurrentViewTypeProvider from 'views/components/views/CurrentViewTypeProvider';
import IfSearch from 'views/components/search/IfSearch';
import IfInteractive from 'views/components/dashboard/IfInteractive';
import HighlightMessageInQuery from 'views/components/messagelist/HighlightMessageInQuery';
import { ViewMetadataStore } from 'views/stores/ViewMetadataStore';
import bindSearchParamsFromQuery from 'views/hooks/BindSearchParamsFromQuery';
import { useSyncWithQueryParameters } from 'views/hooks/SyncWithQueryParameters';
import { AdditionalContext } from 'views/logic/ActionContext';
import DefaultFieldTypesProvider from 'views/components/contexts/DefaultFieldTypesProvider';
import InteractiveContext from 'views/components/contexts/InteractiveContext';
import HighlightingRulesProvider from 'views/components/contexts/HighlightingRulesProvider';
import SearchPageLayoutProvider from 'views/components/contexts/SearchPageLayoutProvider';
import usePluginEntities from 'views/logic/usePluginEntities';
import WidgetFocusProvider from 'views/components/contexts/WidgetFocusProvider';
import WidgetFocusContext from 'views/components/contexts/WidgetFocusContext';
import SearchExecutionState from 'views/logic/search/SearchExecutionState';
import { RefluxActions } from 'stores/StoreTypes';

const GridContainer = styled.div<{ interactive: boolean }>(({ interactive }) => {
  return interactive ? css`
    display: flex;
    overflow: auto;
    height: 100%;

    > *:nth-child(2) {
      flex-grow: 1;
    }
  ` : css`
    flex: 1
  `;
});

const SearchArea = styled(PageContentLayout)(() => {
  const { focusedWidget } = useContext(WidgetFocusContext);

  return css`
    ${focusedWidget?.id && css`
      .page-content-grid {
        display: flex;
        flex-direction: column;
        height: 100%;
        width: 100%;

        /* overflow auto is required to display the message table widget height correctly */
        overflow: ${focusedWidget?.id ? 'auto' : 'visible'};
      }
    `}
  `;
});

const ConnectedSidebar = connect(
  Sidebar,
  { viewMetadata: ViewMetadataStore, searches: SearchStore, view: ViewStore },
  ({ viewMetadata, view, searches }) => ({
    viewMetadata,
    viewIsNew: view.isNew,
    queryId: viewMetadata.activeQuery,
    results: searches?.result?.forId(viewMetadata.activeQuery),
  }),
);

type Props = {
  location: Location,
};

const _searchRefreshConditionChain = (searchRefreshHooks: Array<SearchRefreshCondition>, state: SearchRefreshConditionArguments) => {
  if (!searchRefreshHooks || searchRefreshHooks.length === 0) {
    return true;
  }

  return searchRefreshHooks.every((condition: SearchRefreshCondition) => condition(state));
};

const _refreshIfNotUndeclared = (searchRefreshHooks: Array<SearchRefreshCondition>, executionState: SearchExecutionState, setHasErrors: (hasErrors: boolean) => void) => {
  const { view } = ViewStore.getInitialState();

  return SearchMetadataActions.parseSearch(view.search).then((searchMetadata) => {
    if (_searchRefreshConditionChain(searchRefreshHooks, { view, searchMetadata, executionState })) {
      FieldTypesActions.refresh();

      setHasErrors(false);

      return SearchActions.execute(executionState);
    }

    setHasErrors(true);

    return Promise.reject(searchMetadata);
  });
};

const SearchBarWithStatus = WithSearchStatus(SearchBar);
const DashboardSearchBarWithStatus = WithSearchStatus(DashboardSearchBar);

const ViewAdditionalContextProvider = connect(
  AdditionalContext.Provider,
  { view: ViewStore, configs: SearchConfigStore },
  ({ view, configs: { searchesClusterConfig } }) => ({ value: { view: view.view, analysisDisabledFields: searchesClusterConfig?.analysis_disabled_fields } } as { value: object}),
);

ViewAdditionalContextProvider.displayName = 'ViewAdditionalContextProvider';

const useRefreshSearchOn = (actions: Array<RefluxActions<any>>, refresh: () => Promise<any>) => {
  useEffect(() => {
    let storeListenersUnsubscribes = Immutable.List<() => void>();

    refresh().finally(() => {
      storeListenersUnsubscribes = storeListenersUnsubscribes
        .push(SearchActions.refresh.listen(refresh))
        .push(ViewActions.search.completed.listen(refresh));
    });

    // Returning cleanup function used when unmounting
    return () => { storeListenersUnsubscribes.forEach((unsubscribeFunc) => unsubscribeFunc()); };
  }, [refresh]);
};

const useBindSearchParamsFromQuery = (query: { [key: string]: unknown }) => {
  useEffect(() => {
    const { view } = ViewStore.getInitialState();

    bindSearchParamsFromQuery({ view, query, retry: () => Promise.resolve() });
  }, [query]);
};

const Search = ({ location }: Props) => {
  const { pathname, search } = location;
  const query = `${pathname}${search}`;
  const searchRefreshHooks = usePluginEntities('views.hooks.searchRefresh');
  const [hasErrors, setHasErrors] = useState(false);
  const refreshIfNotUndeclared = useCallback(
    () => _refreshIfNotUndeclared(searchRefreshHooks, SearchExecutionStateStore.getInitialState(), setHasErrors),
    [searchRefreshHooks],
  );

  useBindSearchParamsFromQuery(location.query);
  useSyncWithQueryParameters(query);

  useRefreshSearchOn([SearchActions.refresh, ViewActions.search], refreshIfNotUndeclared);

  useEffect(() => {
    SearchConfigActions.refresh();

    StreamsActions.refresh();
  }, []);

  return (
    <WidgetFocusProvider>
      <WidgetFocusContext.Consumer>
        {({ focusedWidget: { focusing: focusingWidget, editing: editingWidget } = { focusing: false, editing: false } }) => (
          <CurrentViewTypeProvider>
            <IfInteractive>
              <IfDashboard>
                <WindowLeaveMessage />
              </IfDashboard>
            </IfInteractive>
            <InteractiveContext.Consumer>
              {(interactive) => (
                <SearchPageLayoutProvider>
                  <DefaultFieldTypesProvider>
                    <ViewAdditionalContextProvider>
                      <HighlightingRulesProvider>
                        <GridContainer id="main-row" interactive={interactive}>
                          <IfInteractive>
                            <ConnectedSidebar>
                              <FieldsOverview />
                            </ConnectedSidebar>
                          </IfInteractive>
                          <SearchArea>
                            <IfInteractive>
                              <HeaderElements />
                              <IfDashboard>
                                {!editingWidget && <DashboardSearchBarWithStatus onExecute={refreshIfNotUndeclared} />}
                              </IfDashboard>
                              <IfSearch>
                                <SearchBarWithStatus onExecute={refreshIfNotUndeclared} />
                              </IfSearch>

                              <QueryBarElements />

                              <IfDashboard>
                                {!focusingWidget && <QueryBar />}
                              </IfDashboard>
                            </IfInteractive>
                            <HighlightMessageInQuery>
                              <SearchResult hasErrors={hasErrors} />
                            </HighlightMessageInQuery>
                          </SearchArea>
                        </GridContainer>
                      </HighlightingRulesProvider>
                    </ViewAdditionalContextProvider>
                  </DefaultFieldTypesProvider>
                </SearchPageLayoutProvider>
              )}
            </InteractiveContext.Consumer>
          </CurrentViewTypeProvider>
        )}
      </WidgetFocusContext.Consumer>
    </WidgetFocusProvider>
  );
};

export default withLocation(Search);
