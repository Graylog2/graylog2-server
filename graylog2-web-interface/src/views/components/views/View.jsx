// @flow strict
import * as React from 'react';
import PropTypes from 'prop-types';
// $FlowFixMe: imports from core need to be fixed in flow
import { Link } from 'react-router';

import Routes from 'routing/Routes';
import connect from 'stores/connect';
import { EntityListItem } from 'components/common';
import StoreProvider from 'injection/StoreProvider';
import UserTimezoneTimestamp from 'views/components/common/UserTimezoneTimestamp';
import withPluginEntities from 'views/logic/withPluginEntities';

const CurrentUserStore = StoreProvider.getStore('CurrentUser');

const formatTitle = (title, id, disabled = false) => (disabled
  ? <h2>{title}</h2>
  : <Link to={Routes.pluginRoute('DASHBOARDS_VIEWID')(id)}>{title}</Link>);

const _OwnerTag = ({ owner, currentUser }) => {
  if (!owner || owner === currentUser.username) {
    return <span>Last saved</span>;
  }

  return <span>Shared by {owner}, last saved</span>;
};

_OwnerTag.propTypes = {
  owner: PropTypes.string.isRequired,
  currentUser: PropTypes.shape({
    username: PropTypes.string.isRequired,
  }).isRequired,
};

const OwnerTag = connect(_OwnerTag, { currentUser: CurrentUserStore }, ({ currentUser }) => ({ currentUser: currentUser.currentUser }));

// eslint-disable-next-line react/prop-types
const Description = ({ description, owner, createdAt }) => (
  <React.Fragment>
    <div>{description || <i>No description given.</i>}</div>
    <div style={{ color: 'darkgray' }}><OwnerTag owner={owner} /> at <UserTimezoneTimestamp dateTime={createdAt} /></div>
  </React.Fragment>
);

type Plugin = {
  name: string,
  url: string,
};

type RequirementsList = { [string]: Plugin };

const missingRequirements = (requires, requirementsProvided): RequirementsList => Object.entries(requires)
  .filter(([require]) => !requirementsProvided.includes(require))
  .reduce((prev, [key, value]) => ({ ...prev, [key]: value }), {});

const isMissingRequirements = (requires, requirementsProvided) => Object.keys(missingRequirements(requires, requirementsProvided)).length > 0;

type RequirementsProps = {
  requires: RequirementsList,
  requirementsProvided: Array<string>,
};

type MissingRequirementProps = {
  plugin: Plugin,
};

const MissingRequirement = ({ plugin }: MissingRequirementProps) => (
  <a href={plugin.url} target="_blank" rel="noopener noreferrer"><strong>{plugin.name}</strong></a>
);

const Requirements = ({ requires, requirementsProvided }: RequirementsProps) => {
  const missing = missingRequirements(requires, requirementsProvided);
  return Object.keys(missing).length > 0
    ? (
      <h5>
        Missing requirement(s): {Object.values(missing)
        // $FlowFixMe: plugin is of type Plugin, not mixed.
        .map((plugin: Plugin) => <MissingRequirement key={plugin.name} plugin={plugin} />)}
      </h5>
    )
    : null;
};

const View = ({ children, id, title, summary, description, owner, requires, createdAt, requirementsProvided }) => (
  <EntityListItem title={formatTitle(title, id, isMissingRequirements(requires, requirementsProvided))}
                  titleSuffix={summary}
                  description={(
                    <React.Fragment>
                      <Description description={description} owner={owner} createdAt={createdAt} />
                      <Requirements requires={requires} requirementsProvided={requirementsProvided} />
                    </React.Fragment>
                  )}
                  actions={children} />
);

View.propTypes = {
  children: PropTypes.node,
  id: PropTypes.string.isRequired,
  title: PropTypes.string,
  summary: PropTypes.string,
  description: PropTypes.string,
  owner: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
  requires: PropTypes.objectOf(PropTypes.shape({
    name: PropTypes.string.isRequired,
    url: PropTypes.string.isRequired,
  })).isRequired,
  requirementsProvided: PropTypes.arrayOf(PropTypes.string).isRequired,
};

View.defaultProps = {
  children: null,
  title: 'Unnamed View',
  summary: null,
  description: null,
};

export default withPluginEntities(View, { requirementsProvided: 'views.requires.provided' });
