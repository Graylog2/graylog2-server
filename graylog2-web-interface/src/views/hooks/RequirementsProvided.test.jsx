// @flow strict
import type { PluginMetadata } from 'views/logic/views/View';
import * as React from 'react';
import { mount } from 'wrappedEnzyme';

import { PluginStore } from 'graylog-web-plugin/plugin';
import View from 'views/logic/views/View';
import Search from '../logic/search/Search';

import RequirementsProvided from './RequirementsProvided';

jest.mock('graylog-web-plugin/plugin', () => ({
  PluginStore: {
    exports: jest.fn(),
  },
}));
jest.mock('views/Constants', () => ({
  viewsPath: '/path/to/views',
}));

const retry = () => Promise.resolve();

describe('RequirementsProvided', () => {
  const plugin: PluginMetadata = {
    name: 'Pandora\'s Box',
    url: 'https://www.graylog.org',
  };
  it('returns resolved promise for empty requirements', () => {
    PluginStore.exports.mockReturnValue([]);

    const view = View.create().toBuilder().requires({}).build();

    return RequirementsProvided({ view, query: {}, retry });
  });
  it('returns resolved promise if all requirements are provided', () => {
    PluginStore.exports.mockReturnValue(['parameters', 'timetravel', 'hyperspeed']);

    const view = View.create()
      .toBuilder()
      .requires({
        parameters: plugin,
        timetravel: plugin,
        hyperspeed: plugin,
      })
      .build();

    return RequirementsProvided({ view, query: {}, retry });
  });
  it('throws Component if not all requirements are provided', (done) => {
    PluginStore.exports.mockReturnValue(['parameters']);

    const view = View.create()
      .toBuilder()
      .createdAt(new Date('2019-05-29T11:24:51.555Z'))
      .search(
        Search.create()
          .toBuilder()
          .id('5cee6c03675ef9df8b0a7bb0')
          .build(),
      )
      .requires({
        parameters: plugin,
        timetravel: plugin,
        hyperspeed: plugin,
      })
      .build();

    RequirementsProvided({ view, query: {}, retry })
      .catch((Component) => {
        const wrapper = mount(<Component />);
        expect(wrapper).toMatchSnapshot();
        done();
      });
  });
});
