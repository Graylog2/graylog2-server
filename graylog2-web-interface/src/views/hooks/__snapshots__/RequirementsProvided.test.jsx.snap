// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequirementsProvided throws Component if not all requirements are provided 1`] = `
<Component>
  <MissingRequirements
    missingRequirements={
      Object {
        "hyperspeed": Object {
          "name": "Pandora's Box",
          "url": "https://www.graylog.org",
        },
        "timetravel": Object {
          "name": "Pandora's Box",
          "url": "https://www.graylog.org",
        },
      }
    }
    view={
      Object {
        "created_at": 2019-05-29T11:24:51.555Z,
        "description": undefined,
        "id": undefined,
        "owner": undefined,
        "properties": undefined,
        "search_id": "5cee6c03675ef9df8b0a7bb0",
        "state": undefined,
        "summary": undefined,
        "title": undefined,
        "type": undefined,
      }
    }
  >
    <Row
      bsClass="row"
      className="content"
      componentClass="div"
    >
      <div
        className="content row"
      >
        <Col
          bsClass="col"
          className="bootstrapHeading"
          componentClass="div"
          md={6}
          mdOffset={3}
        >
          <div
            className="bootstrapHeading col-md-6 col-md-offset-3"
          >
            <h1>
              View: 
              <em />
            </h1>
            <p>
              Unfortunately executing this view is not possible. It uses the following capabilities which are not available:
            </p>
            <ul>
              <li
                key="timetravel"
              >
                <strong>
                  timetravel
                </strong>
                 - included in 
                <a
                  href="https://www.graylog.org"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Pandora's Box
                </a>
              </li>
              <li
                key="hyperspeed"
              >
                <strong>
                  hyperspeed
                </strong>
                 - included in 
                <a
                  href="https://www.graylog.org"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  Pandora's Box
                </a>
              </li>
            </ul>
          </div>
        </Col>
        <Col
          bsClass="col"
          componentClass="div"
          md={1}
          mdOffset={8}
        >
          <div
            className="col-md-1 col-md-offset-8"
          >
            <LinkContainer
              action="push"
              onlyActiveOnIndex={false}
              to="/path/to/views"
            >
              <ForwardRef
                action="push"
                bsStyle="success"
                onClick={[Function]}
                type="submit"
              >
                <Button
                  action="push"
                  bsStyle="success"
                  onClick={[Function]}
                  type="submit"
                >
                  <StyledComponent
                    action="push"
                    bsStyle="success"
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "SIZES": Array [
                          "large",
                          "small",
                          "xsmall",
                        ],
                        "STYLES": Array [
                          "success",
                          "warning",
                          "danger",
                          "info",
                          "default",
                          "primary",
                          "link",
                        ],
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "Button-c9cbmb-0",
                          "isStatic": false,
                          "lastClassName": "jKsGyd",
                          "rules": Array [
                            [Function],
                          ],
                        },
                        "displayName": "Button",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "Button-c9cbmb-0",
                        "target": [Function],
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                    onClick={[Function]}
                    type="submit"
                  >
                    <Button
                      action="push"
                      active={false}
                      block={false}
                      bsClass="btn"
                      bsStyle="success"
                      className="Button-c9cbmb-0 jKsGyd"
                      disabled={false}
                      onClick={[Function]}
                      type="submit"
                    >
                      <button
                        action="push"
                        className="Button-c9cbmb-0 jKsGyd btn btn-success"
                        disabled={false}
                        onClick={[Function]}
                        type="submit"
                      >
                        Back
                      </button>
                    </Button>
                  </StyledComponent>
                </Button>
              </ForwardRef>
            </LinkContainer>
          </div>
        </Col>
      </div>
    </Row>
  </MissingRequirements>
</Component>
`;
